packages/block-library/src/navigation/edit/index.js

```diff
@@ -98,13 +99,49 @@ function Navigation( {
 
 	const ref = attributes.ref;
 
-	const setRef = ( postId ) => {
-		setAttributes( { ref: postId } );
+	const setRef = ( postSlug ) => {
+		setAttributes( { ref: postSlug } );
 	};
 
 	const recursionId = `navigationMenu/${ ref }`;
 	const hasAlreadyRendered = useHasRecursion( recursionId );
 
+	const generatedTemplatePartHierarchySlug = useSelect(
+		( select ) => {
+			// Use the lack of a clientId as an opportunity to bypass the rest
+			// of this hook.
+			if ( ! clientId ) {
+				return;
+			}
+
+			const { getBlock, getBlockParentsByBlockName } =
+				select( blockEditorStore );
+
+			const withAscendingResults = true;
+			const parentTemplatePartClientIds = getBlockParentsByBlockName(
+				clientId,
+				'core/template-part',
+				withAscendingResults
+			);
+
+			let result = parentTemplatePartClientIds.map(
+				( templatePartClientId ) => {
+					return getBlock( templatePartClientId )?.attributes?.slug;
+				}
+			);
+
+			// Inclusion of this hardcoded string encodes the template part hierarchy
+			// directly into the generated slug. This may be used later to retrieve
+			// the template part hierarchy from the slug.
+			const TEMPLATE_PART_BOUNDARY_DELINEATOR = '-wptp-';
```

Comment:
Before someone says "this is a bad idea" I'm already coming to the conclusion that it is 😄 

Let's discuss the alternatives for how we can preserve the template hierarchy encoded within the slug as WP seems to filter what can be stored here.

Reply:
When the theme is switched:

- say, a visitor visits homepage
- the server renders the new theme's, say, header
- in the header there is an empty navigation block
- at this point rendering _this empty block_, we don't have a slug
- so we need to figure out where this block is (how?)
- build a query
- and ask "Which ones of the saved menus sounds like this query?"
- then retrieve it and render it

It's the same in the editor, when the editor is open following a theme switch _we have an empty navigation block_ so we need to build the hierarchy from block tree (this is easy) and then:

- build a query
- and ask "Which ones of the saved menus sounds like this query?"
- then retrieve it and render it
- make the template part dirty OR use it as a fallback

Reply:
As you say, when you switch Themes the Navigation block in the new Theme has **no information at all**. It doesn't have a `ref` it's just sitting there waiting to be used.

So as you indicate we need to find a way of building a "query" which takes into account where a Navigation block resides within a template We need to be able to do this both within and outside of the Editor.

Let's say we could build a query such as being nested under the following hierarchy:

```
- header
--header-utility
--- core/navigation
```

Our "query" data that we send to the API endpoint becomes something like:

```
['header','header-utility']
```

Now on the API side we need to find the most suitable `wp_navigation` post based on this query. How do we do that?

Well we need to try and match the template hierarchy provided in the query to a Navigation Menu. 

That implies that we need each Navigation Menu to be saved with a `slug` which represents the template hierarchy where it was _created_.

If we don't have that then it's impossible to have a Navigation Menu created in Theme A, map appropriately to a similar (but not necessarily identical) location in Theme B.

We need to preserve the template hierarchy for all Navigation Posts when they are created. The way I'm suggesting we do this is to "encode" it within the slug.

With that in place we can then use the "query" derived from the new Theme to match for Navigation Posts created in other Theme's.

For example, a standard `WP_Query` using the `name` argument might work:

```php
$args = array(
    'name'                   => $query, // this would be "query" derived from the location of the Nav block in the new Theme
    'post_type'              => 'wp_navigation',
);

// Query for the Navigation Post by slug (post_name).
$query = new WP_Query( $args );
```

It's likely that a bare bones `WP_Query` won't cut the mustard, but it gives an example of what I'm trying to achieve.

Reply:
@draganescu and I discussed this away from Github and we came to the conclusion that it's unlikely we will need to decode the template part hierarchy from the slug. Therefore we won't need a delineator in the string.

@draganescu suggested perhaps using something like PHP `levenshtein()` function to determine the similarity of the "query" sent to the API endpoint and the slugs (`post_name`) of each of the Navigation posts. This can be used to determine which one is most similar.

@mtias has also previously suggested looking into the following as possible sources of inspiration:

- https://developer.wordpress.org/reference/functions/wp_old_slug_redirect/
- https://developer.wordpress.org/reference/functions/redirect_guess_404_permalink/
- https://developer.wordpress.org/reference/functions/redirect_canonical/

For now however, we don't need to solve this problem. Indeed it would overcomplicate this PR.